test/integration/builder/selects.js

```diff
@@ -178,6 +179,38 @@ module.exports = function(knex) {
       });
     });
 
+    it('emits error on the stream, if not passed a function, and connecting fails', function() {
+      var expected = new Error();
+      var original = Runner.prototype.ensureConnection;
+      Runner.prototype.ensureConnection = function() {
+        return Promise.reject(expected);
+      };
+
+      var restore = () => {
+        Runner.prototype.ensureConnection = original;
+      };
+
+      var promise = new Promise((resolve, reject) => {
+        var timeout = setTimeout(() => {
+          reject(new Error('Timeout'));
+        }, 5000);
+
+        var stream = knex('accounts').stream();
+        stream.on('error', function(actual) {
+          clearTimeout(timeout);
+
+          if (actual === expected) {
+            resolve();
+          } else {
+            reject(new Error('Stream emitted unexpected error'));
+          }
+        });
+      });
+
+      promise.then(restore, restore);
```

Comment:
doesn't this catch rejected promise, so even if promise is rejected test will pass?

Reply:
No, since the original promise is returned on the next line. This is a more manual [`Promise.prototype.finally`](https://github.com/tc39/proposal-promise-finally).

Reply:
Ok, so test just returns slightly before function is restored. I think it is fine though. Very unlikely to cause any problems.  Is there reason not to use `.finally` ?

Reply:
The sequencing is such that the restore happens before the test finishes, because Mocha will attach the callbacks to the returned promise *after* the restore callbacks.

`.finally` isn't supported yet in any Node.js version.